rbs/CommentsAssociator.cc

```diff
@@ -0,0 +1,282 @@
+           (parser::isa_node<parser::DefMethod>(send->args[0].get()) ||
+            parser::isa_node<parser::DefS>(send->args[0].get())) &&
+           (send->method == core::Names::private_() || send->method == core::Names::protected_() ||
+            send->method == core::Names::public_() || send->method == core::Names::privateClassMethod() ||
+            send->method == core::Names::publicClassMethod() || send->method == core::Names::packagePrivate() ||
+            send->method == core::Names::packagePrivateClassMethod());
+}
+
+bool isAttrAccessorSend(const parser::Send *send) {
+    return (send->receiver == nullptr || parser::isa_node<parser::Self>(send->receiver.get())) &&
+           (send->method == core::Names::attrReader() || send->method == core::Names::attrWriter() ||
+            send->method == core::Names::attrAccessor());
+}
+}; // namespace
+
+void CommentsAssociator::consumePrecedingComments(parser::Node *node, const string_view &prefix) {
+    auto nodeStartLine = core::Loc::pos2Detail(ctx.file.data(ctx), node->loc.beginPos()).line;
+
+    for (auto it = commentByLine.begin(); it != commentByLine.end();) {
+        if (it->first < nodeStartLine) {
+            if (absl::StartsWith(it->second.string, prefix)) {
+                if (auto e = ctx.beginError(it->second.loc, core::errors::Rewriter::RBSUnusedComment)) {
+                    e.setHeader("Unused RBS signature comment. No method definition found after it");
+                }
+            }
+            it = commentByLine.erase(it);
+        } else {
+            break;
+        }
+    }
+}
+
+void CommentsAssociator::consumeDanglingComments(parser::Node *node) {
+    auto nodeEndLine = core::Loc::pos2Detail(ctx.file.data(ctx), node->loc.endPos()).line;
+
+    for (auto it = commentByLine.begin(); it != commentByLine.end();) {
+        if (it->first < nodeEndLine) {
+            if (absl::StartsWith(it->second.string, RBS_PREFIX)) {
+                if (auto e = ctx.beginError(it->second.loc, core::errors::Rewriter::RBSUnusedComment)) {
+                    e.setHeader("Unused RBS signature comment. No method definition found after it");
+                }
+            }
+            it = commentByLine.erase(it);
+        } else {
+            break;
+        }
+    }
+}
+
+void CommentsAssociator::associateCommentsToLines() {
+    for (auto &[start, end] : commentLocations) {
+        auto comment_string = ctx.file.data(ctx).source().substr(start, end - start);
+        auto start32 = static_cast<uint32_t>(start);
+        auto end32 = static_cast<uint32_t>(end);
+        auto comment = CommentNode{core::LocOffsets{start32, end32}, comment_string};
+
+        auto line = core::Loc::pos2Detail(ctx.file.data(ctx), start32).line;
+        commentByLine[line] = comment;
+    }
+}
+
+void CommentsAssociator::associateCommentsToNode(parser::Node *node,
+                                                 const InlinedVector<std::string_view, 2> &prefixes) {
```

Comment:
I don't think we need this complexity, can't we always associate the `#:`, `#|` and `# @`?